src/QirTools/pyqir/qirlib/src/emit/qir/instructions.rs

```diff
@@ -0,0 +1,167 @@
+};
+use crate::{emit::Context, interop::Instruction};
+use inkwell::values::{BasicValueEnum, FunctionValue};
+use std::collections::BTreeMap;
+
+fn get_qubit<'ctx>(
+    name: &String,
+    qubits: &BTreeMap<String, BasicValueEnum<'ctx>>,
+) -> BasicValueEnum<'ctx> {
+    qubits.get(name).unwrap().to_owned()
+}
+
+fn get_register<'ctx>(
+    name: &String,
+    registers: &BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)>,
+) -> (BasicValueEnum<'ctx>, Option<u64>) {
+    registers.get(name).unwrap().to_owned()
+}
+
+pub(crate) fn emit<'ctx>(
+    context: &Context<'ctx>,
+    inst: &Instruction,
+    qubits: &BTreeMap<String, BasicValueEnum<'ctx>>,
+    registers: &BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)>,
+) -> () {
+    let intrinsics = &context.intrinsics;
+    let find_qubit = |name| get_qubit(name, qubits);
+    let ctl = |value| create_ctl_wrapper(context, value);
+    match inst {
+        Instruction::Cx(inst) => {
+            let control = ctl(&find_qubit(&inst.control));
+            let qubit = find_qubit(&inst.target);
+            controlled(context, intrinsics.x_ctl, control, qubit);
+        }
+        Instruction::Cz(inst) => {
+            let control = ctl(&find_qubit(&inst.control));
+            let qubit = find_qubit(&inst.target);
+            controlled(context, intrinsics.z_ctl, control, qubit);
+        }
+        Instruction::H(inst) => {
+            calls::emit_void_call(context, intrinsics.h, &[find_qubit(&inst.qubit)])
+        }
+        Instruction::M { qubit, target } => {
+            measure(context, qubit, target, qubits, registers);
+        }
+        Instruction::Reset(inst) => {
+            calls::emit_void_call(context, intrinsics.reset, &[find_qubit(&inst.qubit)])
+        }
+        Instruction::Rx(inst) => calls::emit_void_call(
+            context,
+            intrinsics.r_x,
+            &[
+                basic_values::f64_to_f64(context, &inst.theta),
+                find_qubit(&inst.qubit),
+            ],
+        ),
+        Instruction::Ry(inst) => calls::emit_void_call(
+            context,
+            intrinsics.r_y,
+            &[
+                basic_values::f64_to_f64(context, &inst.theta),
+                find_qubit(&inst.qubit),
+            ],
+        ),
+        Instruction::Rz(inst) => calls::emit_void_call(
+            context,
+            intrinsics.r_z,
+            &[
+                basic_values::f64_to_f64(context, &inst.theta),
+                find_qubit(&inst.qubit),
+            ],
+        ),
+        Instruction::S(inst) => {
+            calls::emit_void_call(context, intrinsics.s, &[find_qubit(&inst.qubit)])
+        }
+        Instruction::Sdg(inst) => {
+            calls::emit_void_call(context, intrinsics.s_adj, &[find_qubit(&inst.qubit)])
+        }
+        Instruction::T(inst) => {
+            calls::emit_void_call(context, intrinsics.t, &[find_qubit(&inst.qubit)])
+        }
+        Instruction::Tdg(inst) => {
```

Comment:
Elsewhere you've called this `t_adj` or `TAdj`, such that it may be good to use that terminology here as well. In particular, "dagger" is the notation used for the adjoint (e.g.: 𝑇⁺ is the adjoint of 𝑇), so using "adjoint" in preference to "dagger" is more explicit about the meaning here.